In [2]:
# Competition-Solution/notebooks/train/00_data_profiling/00_data_profiling_basic.ipynb

### Data Profiling - Training Dataset

The notebooks within the **'00_data_profiling'** folder perform initial exploratory data analysis on the training dataset, focusing on the three subtasks:

- **Subtask 1:** Call2Action

- **Subtask 2:** Attacks on the Democratic Basic Order (FDGO)

- **Subtask 3:** Violence Detection


Data Profiling is performed in order to get a better understanding of the data we are working with. When data profiling, we are looking to gather a high-level overview and statistical summary of our dataset.
In essence, it is a report-card for our data.

The insights from this data "report-card" will help guide the approach we take for each subtask. 
Additionally, it will provide insights for retrospectively analyzing our models' performance.

---

##### <b>Imports</b>

In [3]:
import sys
import pandas as pd
from rich.console import Console

sys.path.append("../../../src")
import data_utils

console = Console()

##### <b>Loading the data</b>

In [4]:
df_call2action, df_dbo, df_violence = data_utils.load_competition_data(data_dir="../../../data/raw/", type="train")

Loading competition data...

Data of type 'train' loaded successfully.

##### <b>Initial Head and Tail</b>

* ##### C2A Head & Tail

In [5]:
df_call2action.head()

,id,description,C2A
0,874458912592533,"Wenn Du Wert drauf legst, dann erwarten wir di...",False
1,855617527810005,"Macht was ihr wollt, aber schreibt nicht ""Wir ...",False
2,1111407378897684,"Tja, wohl etwas schwierig, jetzt das Geld per ...",False
3,945847662120324,"UInd das Glaube ich nicht,,,in Hotels in Arfri...",False
4,993541110684312,und kein TTIP,False


In [6]:
df_call2action.tail()

,id,description,C2A
6835,854995914538833,Bloß gut die Gegenseite wartet dieses mal nich...,False
6836,1095902123781543,Ekelhaftes Pack!,False
6837,946356702069420,Immer rein mit den is Kämpfern,False
6838,1065799833458439,Gibt es schon,False
6839,985366404835116,Ihr müsst euch erweitern,True


* ##### DBO Head & Tail

In [7]:
df_dbo.head()

,id,description,DBO
0,855025171202574,"Der Riese ist geweckt,mit oder ohne Verräter u...",nothing
1,874452862593138,Gut Ding will Weile haben... (y),nothing
2,982690558436034,Sollen sie doch nach Saudi Arabien,nothing
3,1010187762352980,Volle Zustimmung.??,nothing
4,876306965741061,Mal sehen wann wir an der Erderwärmung schuld ...,nothing


In [8]:
df_dbo.tail()

,id,description,DBO
7449,882699391768485,Demokratie und Gleichberechtigung wirst du in ...,criticism
7450,982653845106372,Ist doch nicht anders zu erwarten erst kamen 1...,nothing
7451,982666798438410,Danke an Ungarn und Respekt. .. genau so! !! ...,nothing
7452,982662171772206,Dafür machen die scheiß Kroaten freie Bahn. Ba...,nothing
7453,1065941433444279,Mehr politische Schulbildung - natürlich- voll...,criticism


* ##### VIO Head & Tail

In [9]:
df_violence.head()

,id,description,VIO
0,951901654848258,"Karl Valentin hat es auf den Punkt gebracht: ""...",False
1,925434340828323,Österreich Herbert Klickl FPÖ zu Lage der Zuwa...,False
2,882849821753442,Es war doch sicher wieder der Nazi von nebenan...,False
3,862798533758571,Es ist egal wieviel am Montag kommen wir bring...,False
4,982677185104038,Ramon De la Perez...Lass deine Krankheit bitte...,False


In [10]:
df_violence.tail()

,id,description,VIO
7778,962236590481431,""" Niemand darf wegen seines Geschlechtes, sein...",False
7779,1063421797029576,Seine Mutti ärgert sich warscheinlich auch übe...,False
7780,1105469482824807,Wer bezahlt das. Wenn sie Geld haben um Mosche...,False
7781,982722861766137,Wenn das so weiter geht mit die Flüchtlinge wi...,False
7782,982767295095027,"schaut mal, dieser kleine süße arme Fratz komm...",False


##### <b>Row (Samples) and Column (Features) Counts</b>

In [11]:
print(f"Subtask 1 (Call2Action): {df_call2action.shape[0]} rows, {df_call2action.shape[1]} columns")
print(f"Subtask 2 (DBO): {df_dbo.shape[0]} rows, {df_dbo.shape[1]} columns")
print(f"Subtask 3 (Violence): {df_violence.shape[0]} rows, {df_violence.shape[1]} columns")

total_rows = df_call2action.shape[0] + df_dbo.shape[0] + df_violence.shape[0]
print(f"Total rows: {total_rows}")

Subtask 1 (Call2Action): 6840 rows, 3 columns
Subtask 2 (DBO): 7454 rows, 3 columns
Subtask 3 (Violence): 7783 rows, 3 columns
Total rows: 22077


<u><p id="row-column-counts">Interpretation</p></u>
- Call2Action has 6,840 rows and three columns<br>
  * Call2Action has 2 classes (TRUE, FALSE)
- DBO has 7,454 rows and three columns<br>
  * DBO has 4 classes (nothing, criticism, agitation, subversive)
- Violence has 7,783 rows and three columns<br>
  * Violence has 2 classes (TRUE, FALSE)

- Total rows: 22,077

*Observation:* 7x larger training dataset compared to the trial dataset (22,077 vs. 3,155 rows). Structure remains the same.

##### <b>Data Types</b>

In [12]:
print(f"Subtask 1 (Call2Action): {df_call2action.dtypes}")
print(f"Subtask 2 (DBO): {df_dbo.dtypes}")
print(f"Subtask 3 (Violence): {df_violence.dtypes}")

Subtask 1 (Call2Action): id              int64
description    object
C2A              bool
dtype: object
Subtask 2 (DBO): id              int64
description    object
DBO            object
dtype: object
Subtask 3 (Violence): id              int64
description    object
VIO              bool
dtype: object


<u><p>Interpretation</p></u>
- Subtask 1 (Call2Action):
  - `id`: int64
  - `description`: object
  - `C2A`: bool
- Subtask 2 (DBO):
  - `id`: int64
  - `description`: object
  - `DBO`: object
- Subtask 3 (Violence):
  - `id`: int64
  - `description`: object
  - `VIO`: bool

Note: 'object' in this case represents string data type.

##### <b>Missing/Null Values</b>

In [13]:
print(f"Subtask 1 (Call2Action):")
print(df_call2action.isnull().sum())
print(f"Subtask 2 (DBO):")
print(df_dbo.isnull().sum())
print(f"Subtask 3 (Violence):")
print(df_violence.isnull().sum())

Subtask 1 (Call2Action):
id             0
description    0
C2A            0
dtype: int64
Subtask 2 (DBO):
id             0
description    0
DBO            0
dtype: int64
Subtask 3 (Violence):
id             0
description    0
VIO            0
dtype: int64


<u><p>Interpretation</p></u>
* No missing/null values across all three subtasks!

##### <b>Duplicate Values within the individual subtask datasets</b>

In [14]:
print(f"Subtask 1 (Call2Action):")
print(df_call2action.duplicated().sum())
print(f"Subtask 2 (DBO):")
print(df_dbo.duplicated().sum())
print(f"Subtask 3 (Violence):")
print(df_violence.duplicated().sum())

Subtask 1 (Call2Action):
0
Subtask 2 (DBO):
0
Subtask 3 (Violence):
0


<u><p>Interpretation</p></u>
No duplicate rows within individual subtask datasets!

*Observation:* Each subtask dataset contains unique rows. This means, that even though our training dataset is larger, it is not artificially inflated and provides unique learning signals. (Provided all descriptions are unique too)

##### <b>Overlap between the subtask datasets</b>

In [14]:
datasets_to_compare = [
    ("Call2Action", df_call2action), 
    ("Violence", df_violence), 
    ("DBO", df_dbo)
]

# Calculates the raw overlaps
overlaps = data_utils.calculate_dataset_overlaps(
    datasets=datasets_to_compare,
    column="description",
    show_details=False
)

# Displays metrics
percentage_results = data_utils.display_dataset_overlap_percentages(
    overlaps=overlaps,
    datasets=datasets_to_compare,
    column="description",
    show_pairwise=True
)

Calculating overlaps across 3 datasets: Call2Action, Violence, DBO

Total unique descriptions appearing in at least two datasets: 7447

───────────────────────────── Overlap Percentages for 'description' across 3 datasets ─────────────────────────────

Total overlap: 7447/11870 unique descriptions appear in 2+ datasets (62.74%)

Pairwise percentages:

Call2Action and Violence share 4562 descriptions:

- 4562/6783 from Call2Action (67.26%)

- 4562/7727 from Violence (59.04%)

- 4562/9948 of all unique descriptions between them (45.86%)

Call2Action and DBO share 3689 descriptions:

- 3689/6783 from Call2Action (54.39%)

- 3689/7306 from DBO (50.49%)

- 3689/10400 of all unique descriptions between them (35.47%)

Violence and DBO share 4194 descriptions:

- 4194/7727 from Violence (54.28%)

- 4194/7306 from DBO (57.40%)

- 4194/10839 of all unique descriptions between them (38.69%)

<u><p>Interpretation:</p></u>

The analysis of overlapping tweet descriptions across the three subtask datasets (Call2Action, Violence, and DBO) reveals the following:

*   **Overall Corpus Overlap:**
    *   Out of a total of **11,870 unique descriptions** across all three datasets combined, **7,447 descriptions (62.74%)** appear in *at least two* of these datasets.
    *   *Observation:* This indicates a **significantly higher level of shared content** compared to the trial dataset (9.21% overlap). The training dataset shows load of overlap, suggesting that many tweets are relevant to multiple subtasks at the same time.

*   **Pairwise Overlap Details:**

    *   **Call2Action vs. Violence:**
        *   Share **4,562** common descriptions.
        *   These 4,562 descriptions represent **67.25%** of Call2Action's unique descriptions and **59.04%** of Violence's unique descriptions.
        *   *Observation:* Strong overlap indicating that many tweets calling for action also contain violence-related content.

    *   **Call2Action vs. DBO:**
        *   Share **3,689** common descriptions.
        *   These 3,689 descriptions represent **54.37%** of Call2Action's unique descriptions and **50.51%** of DBO's unique descriptions.
        *   *Observation:* Substantial overlap showing that calls to action frequently involve attacks on democratic institutions.

    *   **Violence vs. DBO:**
        *   Share **4,194** common descriptions.
        *   These 4,194 descriptions represent **54.29%** of Violence's unique descriptions and **57.42%** of DBO's unique descriptions.
        *   *Observation:* High overlap indicating that violence-related content often targets democratic institutions.

<u><p>Summary & Implications</p></u>

*   *Summary of Observations:*
    *   The training dataset exhibits **dramatically higher overlap** (62.74%) compared to the trial dataset (9.21%), with more than half of each dataset's content being shared with other subtasks.
    *   All pairwise overlaps exceed 50%, indicating strong semantic relationships between the subtasks.
    *   The overlap pattern suggests that harmful content often exhibits multiple characteristics simultaneously.

*   *Implications*
    *   **Multi-Task Learning Potential:** The overlap strongly supports the multi task learning approach that we outlined during the trial dataset analysis and reinforces the fact that all subtasks share an intrinsic relationship with one another.

##### <b>Analysing the overlapping tweets</b>

In [ ]:
df_overlapping_tweets = pd.DataFrame(list(overlaps["total"]), columns=["description"])

console.rule("[bold cyan]Overlapping Tweet Descriptions[/bold cyan]")
console.print(f"Total unique overlapping descriptions: {len(df_overlapping_tweets)}")

# Displays the first 3, last 3, and 3 random samples.
console.print("\n[bold]Head (first 5):[/bold]")
for tweet in df_overlapping_tweets["description"].head(5):
    console.print(f"- \"{tweet}\"")

console.print("\n[bold]Tail (last 5):[/bold]")
for tweet in df_overlapping_tweets["description"].tail(5):
    console.print(f"- \"{tweet}\"")

console.print("\n[bold]Random sample (5):[/bold]")
sample_size = min(5, len(df_overlapping_tweets))
for tweet in df_overlapping_tweets["description"].sample(sample_size, random_state=42):
    console.print(f"- \"{tweet}\"")

───────────────────────────────────────── Overlapping Tweet Descriptions ──────────────────────────────────────────

Total unique overlapping descriptions: 7447

Head (first 5):

- "Mir wird Pegida in der Suchliste nicht mehr als Treffer angezeigt? Weiß jemand was das ist?"

- "Ziel des staates ist es, die deutsche nation auszurotten! Und das klappt allmählich hervorragend."

- "(...)  viele Journalisten sind ob ihrer Islam-Vergötterung anscheinend in einen Blutrausch verfallen und 
bezichtigen jeden Tierschützer, der sich zu Recht über den Verkauf dieser betäubungslos (oder gelähmten) 
geschlachteten Tiere aufregt als „Nazi“ und „Fremdenfeind“. 

Die Islam-Strategen können sich glücklich schätzen, dass sich so viele „ungläubige“ Trottel freiwillig zum Dienst 
melden, die dann jegliche Kritik an den Islam als fremdenfeindlich titulieren, nur um unsere Traditionen und unsere
Kultur widerstandslos in den Wüstensand treten zu können.

Pfui Teufel!

http://journalistenwatch.com/cms/alle-tierfreunde-jetzt-auch-nazis/"

- "Heike Domovski"

- "Thomas Biegel schau in dein postfach unter sonstiges ..."

Tail (last 5):

- "Der Theaterplatz wird doch zur Stunde von Asylbewerbern mit einem Protestcamp belagert. Aufruf an die 
Bevölkerung: spendet Decken, Isomatten, warme Kleidung etc. Wogegen protestieren sie denn eigentlich? Für bessere 
Unterkünfte in Gemeinden und Städten, die mit der riesigen Zahl einfach hoffnungslos überfordert sind? Für 
schnellere Bearbeitung ihrer Asylanträge (wofür wir ja auch auftreten), wo die entsprechenden zuständigen Behörden 
hoffnungslos unterbesetzt sind? Für schnellere Erteilung einer Arbeitserlaubnis? Ich wäre dafür, dass sie auch für 
allgemein nützliche Tätigkeiten mit herangezogen werden. Zum Beispiel für Ordnung und Sauberkeit in 
Asylantenunterkünften und den Anlagen rundherum sorgen. So wie Alg2 Empfänger in Wohngebieten. Reguläre 
Arbeitserlaubnis nach Asylbewilligung. Oder protestieren sie gegen die langen Rückführungszeiten nach Ablehnung des
Asylantrages? Ich bin ja geneigt, noch paar warme Klamotten hin zu bringen, aus reiner Menschlichkeit und 
Nächstenliebe. Aber wenn ich wüsste, dass es ankommt,  würde ich es lieber einem Menschen spenden, der gerade in 
einem libanesischen Flüchtlingslager hockt. Denen geht's viel beschi......, als den Menschen, die bis Deutschland 
gekommen sind."

- "Ist schon gut die Erklärung des Polizeipräsidenten! Ich nehme an, er hat diese in Richtung der gewalttätigen 
Dresden Nazifreien abgegeben! Denn von Pegida geht keine Gewalt aus!"

- "Der Ton geht nicht"

- "Mein tochter hat einen kennegelernt er hat jetzt auch zwei blaue augen :D"

- "Haben diese Hirnlosen BA Versager eigentlich vergessen,  dass wir offizielle 3 Mio. Arbeitslose in Deutschland 
haben. Da sind mit Sicherheit auch genügend Fachkräfte dabei. Und das ist nur die offizielle Zahl an Arbeitslosen. 
Denn die 1,00 Euro Joper und die Zeitarbeiter werden da ja nicht gezählt. Jetzt weiß ich auch,  warum die BRD GmbH 
die Zeitarbeitsfirmen, die in Deutschland wie Pilze aus dem Boden schießen,  so unterstützt. Da können die besser 
die hohen Arbeislosenzahlen  in den Keller drücken. Wenn wir nicht aufpassen,  sind wir Bald Ausländer in unserem 
eigenen Land."

Random sample (5):

- "Ganz Europa leidet, daran ist allein unsere Regierung und die Eu Vollpfosten Schuld!"

- "Ja, hast recht, aber alles fängt klein an und nur auf Fleisch verzichten hilft nicht die Bedingungen in den 
Schweineställen zu verbessern, das führt nur zu noch härterem Wettbewerb und noch schlimmeren Zuständen. Demokratie
ist gut, aber zu wenig Staat lässt Menschen zu Tieren werden. Wir benötigen einen Staat der offen und ehrlich 
Probleme auch beim Namen nennt und nicht geheim irgendwo irgendwas verhandelt. Der Staat der Geheimnisse hat ist 
unehrlich, undemokratisch und nicht haltbar! Erst wenn man sieht wer was gemacht hat wird man sehen was man tun 
kann um die Zustände zu verbessern. Es müssen also zunächst die Karten umgedreht werden."

- "Sie haben noch den Schande- Maas vergessen!"

- "Ich kenne einige Iraker! Arbeiten sehr fleißig. Kommen aber aus einer Familie im Irak, die ihre Kinder studieren
lassen konnten. Was aber jetzt kommt sind definitiv keine Facharbeiter!"

- "Mein Brief an Hofreiter vom 24.02.2016  
___________________________________________________________________________________________________________________
______________________________________                                                                             
                                                        Herr Hofreiter.  
___________________________________________________Es ist mir klar , das sie es absolut nicht Vertragen können das 
sie einfach nur Bedeutungslos für unser Land sind. Die Hetzrede die Sie da heute wieder im Bundestag verzapft haben
, ist an Menschenverachtung und Lügen nicht mehr zu überbieten. Das was sie und ihre Minderheiten Partei sich im 
Deutschen Bundestag leisten , verschlägt einem echt die Sprache. Wann kommt denn die Aufforderung ihrerseits "kauft
nicht mehr bei AfD lern " ? Und wann werden sie die Kennzeichnungspflicht für AfD Parteimitglieder in der 
Öffentlichkeit verlangen? Wann werden die ersten Parteiprograme der AfD in der Öffentlichkeit auf Marktplätzen 
verbrannt ? Wann beginnt ihre Haus und Hof Schlägertruppe "Antifa " Häuser und Wohnorte von AfD Mitgliedern zu 
kennzeichnen? Wann werden die ersten PEGIDA Demonstranten interniert werden nur weil sie nicht ihrer verqueren 
politischen Meinung sind ? Wann werden die ersten Schauprozesse gegen AFD Mitglieder und PEGIDA Demonstranten 
stattfinden? Sie hetzen in unsäglicher Art und Weise gegen Menschen nur weil diese sich nicht ihrem Minderheiten 
Diktat beugen wollen. Sie sind ein geistiger Brandstifter , der mit seiner gefährlichen Rhetorik dazu beiträgt 
Linksradikalen Bodensatz zu immer neuen Gewalttaten zu motivieren und in ihrer radikal gewalttätigen Anschauung 
gegenüber allem andersdenkenden zu bestätigen. Pfui Herr Hofreiter"

<u><p>Interpretation:</p></u>

The overlapping tweets in the training dataset consist of substantive and meaningful content. This has not changed from the trial dataset.

*   **Scale:** With **7,447 overlapping descriptions** (compared to 264 in the trial dataset), the training set presents a significant challenge for data leakage prevention and will ultimately result in less learning signals for an MTL model than initially anticipated or hoped for - after cleaning.

*   **Variable Lengths:** In the case of the training dataset, it becomes apparent, that the tweets have very differing lengths. They range from being quite short to being longer snippets of text, which will likely have implications on training and performance.

##### <b>Basic Statistics: <i>Count, Unique, Top, Freq</i></b>

In [16]:
print(f"Subtask 1 (Call2Action):")
print(df_call2action["description"].describe())
print(f"Subtask 2 (DBO):")
print(df_dbo["description"].describe())
print(f"Subtask 3 (Violence):")
print(df_violence["description"].describe())

Subtask 1 (Call2Action):
count     6840
unique    6783
top         ??
freq         9
Name: description, dtype: object
Subtask 2 (DBO):
count     7454
unique    7306
top         ??
freq        12
Name: description, dtype: object
Subtask 3 (Violence):
count     7783
unique    7727
top         ??
freq        10
Name: description, dtype: object


<u><p>Interpretation</p></u>

*   **`count` (Number of Non-Missing Descriptions):**
    *   **C2A:** 6,840 description field entries.
    *   **DBO:** 7,454 description field entries.
    *   **VIO:** 7,783 description field entries.
    *   *Observation:* Matches the total row counts for each subtask, confirming no missing tweet descriptions.

*   **`unique` (Number of Distinct Descriptions):**
    *   **C2A:** 6,783 unique tweet texts.
        *   *Observation:* 57 duplicate tweets (0.83% duplication rate).
    *   **DBO:** 7,306 unique tweet texts.
        *   *Observation:* 148 duplicate tweets (1.99% duplication rate).
    *   **VIO:** 7,727 unique tweet texts.
        *   *Observation:* 56 duplicate tweets (0.72% duplication rate).

*   **`top` (Most Frequent Description):**
    *   All subtasks show "??" as the most frequent description, indicating short, potentially uninformative tweets. (Noise? Likely the same tweet in different subtask datasets?)

*   **`freq` (Frequency of the Top Description):**
    *   **C2A:** Most frequent tweet appears 9 times.
    *   **DBO:** Most frequent tweet appears 12 times.
    *   **VIO:** Most frequent tweet appears 10 times.
    *   *Observation:* Higher frequency counts compared to trial dataset, but they still only represent a small fraction of total data.

<u><p>Summary & Implication</p></u>

*   *Observation:* The training datasets show low levels of text duplication (0.72-1.99%) with the majority of descriptions being unique. The duplication rates are slightly higher than in the trial dataset. This shows that even though we have unique **rows** of data, the learning signals are not unique.
*   *Implication:* The dataset has good data quality and diversity. The small percentage of tweet description duplicates is unlikely to have a major impact on model performance but we will ensure that the validation set does not contain any duplicates of the training set anyways, during our training dataset preprocessing. Good practice and keeps our evaluation metrics honest.